File: broker.go/internal/stores/sqlite/queue_store.go
package sqlite

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"monstermq.io/edge/internal/stores"
)

const queueTable = "messagequeue"

type QueueStore struct {
	db                *DB
	visibilityTimeout time.Duration
}

func NewQueueStore(db *DB, visibilityTimeout time.Duration) *QueueStore {
	if visibilityTimeout <= 0 {
		visibilityTimeout = 30 * time.Second
	}
	return &QueueStore{db: db, visibilityTimeout: visibilityTimeout}
}

func (q *QueueStore) Close() error { return nil }

func (q *QueueStore) EnsureTable(ctx context.Context) error {
	stmts := []string{
		`CREATE TABLE IF NOT EXISTS ` + queueTable + ` (
            msg_id INTEGER PRIMARY KEY AUTOINCREMENT,
            message_uuid TEXT NOT NULL,
            client_id TEXT NOT NULL,
            topic TEXT NOT NULL,
            payload BLOB,
            qos INTEGER NOT NULL,
            retained INTEGER NOT NULL DEFAULT 0,
            publisher_id TEXT,
            creation_time INTEGER NOT NULL,
            message_expiry_interval INTEGER,
            vt INTEGER NOT NULL DEFAULT (strftime('%s', 'now')),
            read_ct INTEGER NOT NULL DEFAULT 0
        )`,
		`CREATE INDEX IF NOT EXISTS ` + queueTable + `_fetch_idx ON ` + queueTable + ` (client_id, vt)`,
		`CREATE INDEX IF NOT EXISTS ` + queueTable + `_client_uuid_idx ON ` + queueTable + ` (client_id, message_uuid)`,
	}
	for _, s := range stmts {
		if _, err := q.db.Exec(s); err != nil {
			return err
		}
	}
	return nil
}

func (q *QueueStore) Enqueue(ctx context.Context, clientID string, msg stores.BrokerMessage) error {
	return q.EnqueueMulti(ctx, msg, []string{clientID})
}

func (q *QueueStore) EnqueueMulti(ctx context.Context, msg stores.BrokerMessage, clientIDs []string) error {
	if len(clientIDs) == 0 {
		return nil
	}
	insert := fmt.Sprintf(`INSERT INTO %s
        (message_uuid, client_id, topic, payload, qos, retained, publisher_id, creation_time, message_expiry_interval)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`, queueTable)
	q.db.Lock()
	defer q.db.Unlock()
	tx, err := q.db.Conn().BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	stmt, err := tx.PrepareContext(ctx, insert)
	if err != nil {
		_ = tx.Rollback()
		return err
	}
	defer stmt.Close()
	for _, cid := range clientIDs {
		retained := 0
		if msg.IsRetain {
			retained = 1
		}
		var expiry sql.NullInt64
		if msg.MessageExpiryInterval != nil {
			expiry = sql.NullInt64{Int64: int64(*msg.MessageExpiryInterval), Valid: true}
		}
		if _, err := stmt.ExecContext(ctx, msg.MessageUUID, cid, msg.TopicName, msg.Payload, int(msg.QoS), retained, msg.ClientID, msg.Time.UnixMilli(), expiry); err != nil {
			_ = tx.Rollback()
			return err
		}
	}
	return tx.Commit()
}

func (q *QueueStore) Dequeue(ctx context.Context, clientID string, batchSize int) ([]stores.BrokerMessage, error) {
	if batchSize <= 0 {
		batchSize = 10
	}
	now := time.Now().Unix()
	newVT := now + int64(q.visibilityTimeout.Seconds())

	q.db.Lock()
	defer q.db.Unlock()
	tx, err := q.db.Conn().BeginTx(ctx, nil)
	if err != nil {
		return nil, err
	}
	rows, err := tx.QueryContext(ctx,
		fmt.Sprintf(`SELECT msg_id, message_uuid, topic, payload, qos, retained, publisher_id, creation_time, message_expiry_interval
                     FROM %s WHERE client_id = ? AND vt <= ? ORDER BY msg_id LIMIT ?`, queueTable),
		clientID, now, batchSize)
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	type rec struct {
		msgID int64
		msg   stores.BrokerMessage
	}
	recs := []rec{}
	for rows.Next() {
		var (
			r        rec
			retained int
			pubID    sql.NullString
			created  int64
			expiry   sql.NullInt64
			qos      int
		)
		if err := rows.Scan(&r.msgID, &r.msg.MessageUUID, &r.msg.TopicName, &r.msg.Payload, &qos, &retained, &pubID, &created, &expiry); err != nil {
			rows.Close()
			_ = tx.Rollback()
			return nil, err
		}
		r.msg.QoS = byte(qos)
		r.msg.IsRetain = retained == 1
		r.msg.IsQueued = true
		r.msg.ClientID = pubID.String
		r.msg.Time = time.UnixMilli(created)
		if expiry.Valid {
			v := uint32(expiry.Int64)
			r.msg.MessageExpiryInterval = &v
		}
		recs = append(recs, r)
	}
	rows.Close()

	if len(recs) > 0 {
		stmt, err := tx.PrepareContext(ctx, fmt.Sprintf("UPDATE %s SET vt = ?, read_ct = read_ct + 1 WHERE msg_id = ?", queueTable))
		if err != nil {
			_ = tx.Rollback()
			return nil, err
		}
		for _, r := range recs {
			if _, err := stmt.ExecContext(ctx, newVT, r.msgID); err != nil {
				stmt.Close()
				_ = tx.Rollback()
				return nil, err
			}
		}
		stmt.Close()
	}
	if err := tx.Commit(); err != nil {
		return nil, err
	}

	out := make([]stores.BrokerMessage, len(recs))
	for i, r := range recs {
		out[i] = r.msg
	}
	return out, nil
}

func (q *QueueStore) Ack(ctx context.Context, clientID, messageUUID string) error {
	_, err := q.db.Exec(fmt.Sprintf("DELETE FROM %s WHERE client_id = ? AND message_uuid = ?", queueTable), clientID, messageUUID)
	return err
}

func (q *QueueStore) PurgeForClient(ctx context.Context, clientID string) (int64, error) {
	res, err := q.db.Exec(fmt.Sprintf("DELETE FROM %s WHERE client_id = ?", queueTable), clientID)
	if err != nil {
		return 0, err
	}
	n, _ := res.RowsAffected()
	return n, nil
}

func (q *QueueStore) Count(ctx context.Context, clientID string) (int64, error) {
	row := q.db.Conn().QueryRowContext(ctx, fmt.Sprintf("SELECT COUNT(*) FROM %s WHERE client_id = ?", queueTable), clientID)
	var n int64
	err := row.Scan(&n)
	return n, err
}

func (q *QueueStore) CountAll(ctx context.Context) (int64, error) {
	row := q.db.Conn().QueryRowContext(ctx, fmt.Sprintf("SELECT COUNT(*) FROM %s", queueTable))
	var n int64
	err := row.Scan(&n)
	return n, err
}

File: broker.rs/src/stores/sqlite/queue_store.rs
use super::db::Db;
use crate::stores::traits::{QueueStore, StoreResult};
use crate::stores::types::BrokerMessage;
use async_trait::async_trait;
use chrono::{TimeZone, Utc};
use sqlx::Row;
use std::time::Duration;

pub struct SqliteQueueStore {
    db: Db,
    visibility_timeout: Duration,
}

impl SqliteQueueStore {
    pub fn new(db: Db, visibility_timeout: Duration) -> Self {
        let visibility_timeout = if visibility_timeout.is_zero() { Duration::from_secs(30) } else { visibility_timeout };
        Self { db, visibility_timeout }
    }
}

#[async_trait]
impl QueueStore for SqliteQueueStore {
    async fn ensure_table(&self) -> StoreResult<()> {
        sqlx::query(r#"CREATE TABLE IF NOT EXISTS messagequeue (
            msg_id INTEGER PRIMARY KEY AUTOINCREMENT,
            message_uuid TEXT NOT NULL,
            client_id TEXT NOT NULL,
            topic TEXT NOT NULL,
            payload BLOB,
            qos INTEGER NOT NULL,
            retained INTEGER NOT NULL DEFAULT 0,
            publisher_id TEXT,
            creation_time INTEGER NOT NULL,
            message_expiry_interval INTEGER,
            vt INTEGER NOT NULL DEFAULT (strftime('%s', 'now')),
            read_ct INTEGER NOT NULL DEFAULT 0
        )"#).execute(&self.db.pool).await?;
        sqlx::query("CREATE INDEX IF NOT EXISTS messagequeue_fetch_idx ON messagequeue (client_id, vt)").execute(&self.db.pool).await?;
        sqlx::query("CREATE INDEX IF NOT EXISTS messagequeue_client_uuid_idx ON messagequeue (client_id, message_uuid)").execute(&self.db.pool).await?;
        Ok(())
    }

    async fn enqueue(&self, client_id: &str, msg: &BrokerMessage) -> StoreResult<()> {
        self.enqueue_multi(msg, &[client_id.to_string()]).await
    }

    async fn enqueue_multi(&self, msg: &BrokerMessage, client_ids: &[String]) -> StoreResult<()> {
        if client_ids.is_empty() { return Ok(()); }
        let mut tx = self.db.pool.begin().await?;
        for cid in client_ids {
            sqlx::query(r#"INSERT INTO messagequeue
                (message_uuid, client_id, topic, payload, qos, retained, publisher_id, creation_time, message_expiry_interval)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"#)
                .bind(&msg.message_uuid).bind(cid)
                .bind(&msg.topic_name).bind(&msg.payload)
                .bind(msg.qos as i64)
                .bind(if msg.is_retain { 1i64 } else { 0 })
                .bind(&msg.client_id)
                .bind(msg.time.timestamp_millis())
                .bind(msg.message_expiry_interval.map(|e| e as i64))
                .execute(&mut *tx).await?;
        }
        tx.commit().await?;
        Ok(())
    }

    async fn dequeue(&self, client_id: &str, batch_size: i64) -> StoreResult<Vec<BrokerMessage>> {
        let limit = if batch_size <= 0 { 10 } else { batch_size };
        let now = Utc::now().timestamp();
        let new_vt = now + self.visibility_timeout.as_secs() as i64;

        let mut tx = self.db.pool.begin().await?;
        let rows = sqlx::query(r#"SELECT msg_id, message_uuid, topic, payload, qos, retained, publisher_id, creation_time, message_expiry_interval
                                FROM messagequeue WHERE client_id = ? AND vt <= ? ORDER BY msg_id LIMIT ?"#)
            .bind(client_id).bind(now).bind(limit).fetch_all(&mut *tx).await?;
        let mut out = Vec::with_capacity(rows.len());
        let mut ids = Vec::with_capacity(rows.len());
        for r in rows {
            let mid: i64 = r.try_get("msg_id")?;
            let muuid: String = r.try_get("message_uuid").unwrap_or_default();
            let topic: String = r.try_get("topic").unwrap_or_default();
            let payload: Vec<u8> = r.try_get("payload").unwrap_or_default();
            let qos: i64 = r.try_get("qos").unwrap_or(0);
            let retained: i64 = r.try_get("retained").unwrap_or(0);
            let pub_id: Option<String> = r.try_get("publisher_id").ok();
            let created: i64 = r.try_get("creation_time").unwrap_or(0);
            let expiry: Option<i64> = r.try_get("message_expiry_interval").ok();
            let mut msg = BrokerMessage {
                message_uuid: muuid,
                topic_name: topic,
                payload,
                qos: qos as u8,
                is_retain: retained == 1,
                is_queued: true,
                client_id: pub_id.unwrap_or_default(),
                time: Utc.timestamp_millis_opt(created).single().unwrap_or_else(Utc::now),
                ..Default::default()
            };
            if let Some(e) = expiry { msg.message_expiry_interval = Some(e as u32); }
            out.push(msg);
            ids.push(mid);
        }
        for id in ids {
            sqlx::query("UPDATE messagequeue SET vt = ?, read_ct = read_ct + 1 WHERE msg_id = ?")
                .bind(new_vt).bind(id).execute(&mut *tx).await?;
        }
        tx.commit().await?;
        Ok(out)
    }

    async fn ack(&self, client_id: &str, message_uuid: &str) -> StoreResult<()> {
        sqlx::query("DELETE FROM messagequeue WHERE client_id = ? AND message_uuid = ?")
            .bind(client_id).bind(message_uuid).execute(&self.db.pool).await?;
        Ok(())
    }

    async fn purge_for_client(&self, client_id: &str) -> StoreResult<i64> {
        let r = sqlx::query("DELETE FROM messagequeue WHERE client_id = ?").bind(client_id).execute(&self.db.pool).await?;
        Ok(r.rows_affected() as i64)
    }

    async fn count(&self, client_id: &str) -> StoreResult<i64> {
        let row = sqlx::query("SELECT COUNT(*) AS c FROM messagequeue WHERE client_id = ?").bind(client_id).fetch_one(&self.db.pool).await?;
        Ok(row.try_get::<i64, _>("c").unwrap_or(0))
    }

    async fn count_all(&self) -> StoreResult<i64> {
        let row = sqlx::query("SELECT COUNT(*) AS c FROM messagequeue").fetch_one(&self.db.pool).await?;
        Ok(row.try_get::<i64, _>("c").unwrap_or(0))
    }
}
